### DISCUSSION TITLE
Using Firebase Authentication (or 3rd party) with Supabase

### DISCUSSION THREAD

**Discussion by miro-ur** [supabase/supabase#391] [Questions] [answered]
Hello,

I have an existing web (Angular) application written using Firebase backend. It utilizes various Firebase modules: Authentication (with custom mailer templates), Firestore with triggers, PubSub, CludFunctions, Storage

I would like to bring Supabase in to the fold and replace the Firestore db with Supabase/Postgres (superior querying, views, RLS, search, extensions, ecosystem, open source) while keeping the rest of my application largely intact to minimize impact and cost and since not all modules with full parity features are available just yet in Supabase. 

What are my options if I wanted to largely keep existing code base but replace Firestore with Supabase ?
Is there a way to bring a 3rd party Auth provider into the mix (Firebase, Auth0, Okta etc ...) ?

Any suggestion or help would be welcome.

**Comment by kiwicopple** [comment, marked as answer]
Hey @miro-ur, since we're "just postgres" you can do this in whichever way you'd usually use.  I can outline the steps below, I just want to check - did you want to use Firebase Auth + Postgres Row Level security? Or will you add your Authorization security in your middleware (eg, you have an API server).

![image](https://user-images.githubusercontent.com/10214025/102585145-1c94f500-4108-11eb-8c47-f0b73fc3a6f1.png)

**Reply by miro-ur** [reply]
Hey @kiwicopple, thanks for the diagram. 
I'm interested in the "no servers" setup. I would like to keep firebase as auth provider and be able to use supabase as datastore - from frontend apps and firebase cloud functions.

I have a small prototype working - had to switch to using postgrest-js instead of supabase-js client because the supabase-js client was giving me trouble on cloud functions ("ReferenceError: globalThis is not defined", caused by the [globalThis.localStorage](https://github.com/supabase/supabase-js/blob/master/src/SupabaseClient.ts#L13))

The idea is to expose a cloud function (secured by firebase auth) to mint my own token using supabase's secret JWT and use this token to then authenticate requests. I would sync user entries (want the same user_id to keep things clean) between systems and keep firebase as my main auth provider (master record). 

I think this might work and should achieve my original goal. It allows me to move some of my data backend from firestore to supabase and start using RLS and all the goodies Postgres has to offer. 

My inspiration was how Firebase allows to authenticate with JWT tokens via [Create Custom Tokens](https://firebase.google.com/docs/auth/admin/create-custom-tokens).

Anything that jumps out at you with this approach ? Any suggestions / things you would have done differently?

**Reply by kiwicopple** [reply]
I think this is exactly what I would have done @miro-ur . Nice work 


Happy new year!

**Reply by mstade** [reply]
Hi!

We're also looking at Supabase and trying to figure out how to do auth. We want to make use of RLS, so presumably when using the client we will need to sign in users. This is somewhat complicated by the fact that we won't have user e-mails, instead we'll be making use of a 2FA solution called BankID, which is essentially an electronic id system developed and backed by all tier 1 banks in Sweden. There are similar solutions in other EU countries, such as NemID, FrejaID, Vipps etc.

We can set up middleware to do the actual eID authentication flow, which would net us some data that we could then use to generate a token compatible with Supabase, that we can then use to effectively impersonate a user in our application, netting us the gain of RLS features etc.

I can't seem to find any documentation on this use case – which is fine – I'd be happy to contribute at least a starting point for documentation. But I wonder, is this use case supported at all? Is there an endpoint we can call to generate a token for impersonating a user (ideal, so any potential vulnerabilities lie with Supabase – more eyeballs and all that!) or can we generate a token in our middleware (not ideal – only our eyeballs on the prize!)

I suppose what I'm asking is – is supabase compatible with OIDC or are there any plans to implement this capability?

**Reply by logemann** [reply]
trying to do the same with Auth0 as external IDP. The tricky part is, i dont want to use an external cloud function to mint the token for supabase and somehow thinking about a way to do everything inside Auth0.

**Reply by offline-first** [reply]
Hello @miro-ur do you have any working example?

**Reply by dawsonbristow20** [reply]
Some documentation for this would be awesome @miro-ur

**Reply by johanneskares** [reply]
Yes, agree, @miro-ur  would be great to know how you make this work

**Reply by ansh** [reply]
@miro-ur If you could share a gist, that would be great!

**Comment by klongmitre** [comment]
@kiwicopple - has there been any advancement on this? A reference implementation perhaps? I'm particularly interested in incorporating Okta as the identity/auth provider and using that with Supabase to handle RLS.

**Comment by swapnil-srivastava** [comment]
@kiwicopple  - same here, is there any advancement over this. As with @miro-ur, I wanted to use the Firebase Authentication + Supabase/postgres RLS for my app or we have to use Supabase/Auth + Supabase/postgres RLS for it to work.

**Comment by josh-burton** [comment]
Would be great to have some guidance on this :)

**Comment by josh-burton** [comment]
I've been thinking about this overnight, and wonder if a solution where, when a user is created in firebase auth, a trigger is used to create a supabase signed jwt with their user id, and this is then stored in a custom claim on the firebase token.

Would that work?

**Reply by brunobely** [reply]
Sounds like a good idea, did you ever attempt this?

**Reply by josh-burton** [reply]
No I didn't. I'm not 100% on the security of this approach. Presumably the generated supabase jwt would need to be invalidated and re-generated at some point.

**Comment by evelant** [comment]
It looks like some people have made it work https://discord.com/channels/839993398554656828/1067498867513376819/1067557342947844137

I'd be interested to hear anybody else's experience doing this. Has it gone smoothly or are there hidden stumbling blocks?

Edit: I made it work without issue. See the post below for details.

**Comment by gregorym** [comment]
I am also curious. I'd like it to use our current Firebase Auth (Google) and create+authenticate a user in supabase.

Would this be possible?

**Reply by evelant** [reply]
Yes. I'm doing it in my project and it works great! Here's the high level of it:

1. Send firebase token from your client to an edge function (I call mine generate-token)
2. Verify the validity of the firebase token using a JWT library or the firebase sdk (I use "jose")
3. Sign a new token with your JWT secret that has the claims you want (usually 'authenticated' and a 'sub' of your firebase uid)
4. Send the new token back to the client and use it to create your Supabase SDK instance
5. Listen for firebase auth token expiry on your client and refresh the auth token as necessary

And here are the details

The `generate-token` edge function
```ts
import { serve } from "https://deno.land/std@0.177.0/http/server.ts"
import "https://deno.land/x/xhr@0.1.1/mod.ts"
import * as jose from "https://deno.land/x/jose@v4.13.1/index.ts"

//@ts-ignore performance is missing from the supabase edge runtime and not documented anywhere...
globalThis.shimPerformance = {
    now: () => new Date().getTime(),
    timeOrigin: 0,
}
//@ts-ignore performance is missing from the supabase edge runtime and not documented anywhere...
globalThis.performance = {
    now: () => new Date().getTime(),
    timeOrigin: 0,
}

let hasInitialized = false
/**
 * call firebase.initializeApp with credentials so we can verify the firebase auth token is legit
 * This is only used in development with the local firebase auth emulator. Jose verifies the tokens in production.
 */
const initFirebaseApp = async () => {
    const firebase = (await import("https://esm.sh/firebase-admin@11.5.0?bundle&target=deno&no-check")).default
    if (hasInitialized) return firebase
    hasInitialized = true
    if (Deno.env.has("FIRESTORE_EMULATOR_HOST") && Deno.env.has("FIREBASE_AUTH_EMULATOR_HOST")) {
        console.log(`initializing firebase app with emulators`)
        firebase.initializeApp({ projectId: Deno.env.get("FIREBASE_PROJECT_ID") })
    } else {
        const serviceAccount = JSON.parse(Deno.env.get("FIREBASE_SERVICE_ACCOUNT") || "{}")
        console.log(`initializing firebase app ${serviceAccount.project_id}`)
        try {
            const credential = firebase.credential.cert(serviceAccount)
            //@ts-ignore types don't work correctly with deno
            const firebaseConfig: firebase.AppOptions = {
                projectId: serviceAccount.project_id,
                credential,
                databaseURL: serviceAccount.database_url,
            }
            firebase.initializeApp(firebaseConfig)
        } catch (err: any) {
            console.error(`Error initializing firebase app`, err)
        }
    }
    return firebase
}

const verifyFirebaseEmulatorToken = async (token: string): Promise<jose.JWTPayload | undefined> => {
    const firebase = await initFirebaseApp()
    try {
        const v = await firebase.auth().verifyIdToken(token)
        return v
    } catch (err: unknown) {
        console.error(`Error verifying firebase emulator token`, err)
    }
    return undefined
}

const verifyFirebaseToken = async (token: string): Promise<jose.JWTPayload | undefined> => {
    try {
        const projectId = Deno.env.get("FIREBASE_PROJECT_ID")
        if (!projectId || !["my-project-id", "my-dev-project-id"].includes(projectId)) {
            throw new Error(`invalid project id ${projectId}`)
        }
        const firebaseHeader = jose.decodeProtectedHeader(token)
        if (!firebaseHeader) {
            console.error(`failed to parse firebase token header`)
            return undefined
        }
        const kid = firebaseHeader.kid

        if (!kid) {
            console.error(`firebase token header missing kid`)
            return undefined
        }

        console.log(`verifying firebase token with header ${JSON.stringify(firebaseHeader, undefined, 2)}...`)
        const JWKS = jose.createRemoteJWKSet(
            new URL("https://www.googleapis.com/service_accounts/v1/jwk/securetoken@system.gserviceaccount.com")
        )

        const { payload, protectedHeader } = await jose.jwtVerify(token, JWKS)

        if (projectId !== payload.aud) {
            throw new Error(`firebase token is valid, but aud mismatches: ${payload.aud} !== ${projectId}`)
        }

        console.log(`verified firebase token`, { payload, protectedHeader })
        return payload
    } catch (error: any) {
        console.error(`Error verifying firebase prod token`, error)
        return undefined
    }
}

/**
 * Given a firebase token, verify it's valid via firebase then issue a supabase token for the same uid
 * This effectively allows firebase auth to be used as a supabase auth provider
 */
export const generateSupabaseToken = async (token: string, firebaseUser: { [key: string]: any }) => {
    console.log(`generateSupabaseToken from firebase token: ${token}`)

    const verifiedToken = Deno.env.has("FIREBASE_AUTH_EMULATOR_HOST")
        ? await verifyFirebaseEmulatorToken(token)
        : await verifyFirebaseToken(token)
    // const verifiedToken = await verifyFirebaseToken(token)
    if (!verifiedToken) {
        console.error(`failed to verify firebase token ${token}`)
        return undefined
    }

    const secret = new TextEncoder().encode(Deno.env.get("JWT_KEY"))
    if (!secret || secret.length === 0) {
        console.error(`missing JWT_KEY to sign key!`)
        return undefined
    }

    const claims = verifiedToken
    if (claims.sub !== firebaseUser.uid) {
        throw new Error(
            `invalid firebase user data, uid mismatch between token and data: ${claims.sub} !== ${
                firebaseUser.uid
            } ${JSON.stringify({ claims, firebaseUser }, undefined, 2)}`
        )
    }
    const jwt = await new jose.SignJWT({
        exp: claims.exp!,
        sub: claims.sub!,
        role: "authenticated",
        user_metadata: null,
        app_metadata: null,
        email: claims.email,
    })
        .setProtectedHeader({ alg: "HS256", typ: "JWT" })
        .setAudience("authenticated")
        .sign(secret)

    console.log(`Generated new supabase jwt: ${jwt}`)
    return jwt
}

serve(async req => {
    const { token, firebaseUser } = await req.json()

    console.log(`got firebaseUser `, firebaseUser)
    const newToken = await generateSupabaseToken(token, firebaseUser)


    const data = {
        token: newToken,
    }

    // I've got some extra code here to insert data into `auth.users`. Not totally necessary, unused at the moment.

    return new Response(JSON.stringify(data), {
        headers: { "Content-Type": "application/json" },
    })
})
```

Then on the client I create my supabase client like so

```ts
const getTokenFromFirebaseToken = async (firebaseToken: string, firebaseUser: FirebaseAuthUser) => {

    try {
        const resp = await axios.post(
            `${creds.fnsUrl}/generate-token`,
            {
                token: firebaseToken,
                firebaseUser,
            },
            {
                headers: {
                    Authorization: `Bearer ${creds.key}`, //supabase anon key
                    "Content-Type": "application/json",
                },
            }
        )

        const { token } = resp.data
        if (!token) {
            throw new Error(
                `failed to get token from firebase token ${firebaseToken} resp ${JSON.stringify(resp.data)}`
            )
        }
        return token
    } catch (e: any) {
        log.error(`failed to get token from firebase token ${e.message}`, e)
        throw e
    }
}


export const setupDbConnection = async (
    localStorageInstance: { getItem: any; setItem: any; removeItem: any },
    firebaseToken: string,
    firebaseUser: FirebaseAuthUser
) => {
    const key = await getTokenFromFirebaseToken(firebaseToken, firebaseUser)

    const authOpts = {
        storage: localStorageInstance,
        autoRefreshToken: false,
        persistSession: false,
        detectSessionInUrl: false,
    }

    const db = createClient<Database>(creds.url, creds.key, {
        auth: authOpts,
        global: {
            headers: {
                Authorization: `Bearer ${key}`, // custom token with 'authenticated' role signed with jwt secret
            },
        },
        realtime: {
            headers: {
                apikey: `Bearer ${key}`, // custom token with 'authenticated' role signed with jwt secret
            },
            params: {
                apikey: creds.key, // supabase anon key
            },
        },
    })

    db.realtime.setAuth(key)
    return db
}

```


Then whenever it's time for token refresh

```ts


/**
 * Hax to set new auth tokens on the supabase client when we generate a new token from a firebase token
 * See these issues
 * https://github.com/supabase/gotrue/issues/957
 * https://github.com/supabase/supabase-js/issues/553
 */
export const refreshSupabaseTokenFromFirebaseToken = async (
    db: MySupabaseClient | null,
    firebaseToken: string,
    firebaseUser: FirebaseAuthUser
) => {
    const key = await getTokenFromFirebaseToken(firebaseToken, firebaseUser)
    if (!db) {
        throw `db not initialized yet`
    }
    //@ts-ignore protected fields
    db.headers.Authorization = `Bearer ${key}`
   

    //@ts-ignore
    db.auth.headers.Authorization = `Bearer ${key}`
    //@ts-ignore
    db.rest.headers.Authorization = `Bearer ${key}`
    db.realtime.setAuth(key)

  
    db.functions.setAuth(key)
  
    return key
}

```


Hope that's helpful!

**Reply by gregorym** [reply]
@evelant Dude, **THANK YOU**!

Let me know if i'm wrong but this does not create an supabase user, right?
it uses a new signed JWT to just authenticate a user?

**Reply by evelant** [reply]
No prob 🙂

 Correct. You don't actually need a "user" in `auth.users` since you're not using Supabase auth. I have a bit of extra code that inserts one anyway, but it doesn't get used for anything (at least for now).

**Reply by gregorym** [reply]
If I use RLS, do you know how I would create a user?

**Reply by evelant** [reply]
You don't need an entry in `auth.users` for RLS. The code above works fine with RLS. Just use `auth.jwt() ->> 'sub'` to get the user id (sub) from your custom signed token.

Like so
```sql
CREATE POLICY "Enable full access to rows owned by authenticated user" ON "public"."activities" TO "authenticated" USING ((("auth"."jwt"() ->> 'sub'::"text") = "userId"));
```

**Reply by subhubirl** [reply]
Hi @evelant this is great content, thanks so much. 
Question: I'm currently using FlutterFlow with Firebase Auth and Supabase as my database but now I need to enable RLS. This looks like exactly what I need to keep my current setup but do you know if this will/can work in FlutterFlow? It would be a dream if I could just take this code and run with it...but I don't know where to invoke it from if it would work

**Reply by subhubirl** [reply]
@evelant @dhurbhat if I am to add this to FlutterFlow via git, where would I create these files and what folders should they be in? 
Any pointers are greatly appreciated, thanks 🙏

**Reply by dhurbhat** [reply]
what is needed is a translation of Firebase token to one that Supabase can mint (create new). 

1. Create a Supabase Edge function Use the server code to create an edge function - give it some url: `https://<yourproject>.supabase.co/functions/v1/mint-token`.  Supabase has good docs on how to setup an environment for developing Edge Functions, how to test these locally and how to deploy to Supabase cloud.
2. `mint-token`copy  Server code to your Edge Functions' `index.ts` - if you want to you can modify the claims
3. In your Flutter (or FlutterFlow app), add a new dart file that does the following:

- Add an ID Token listener to Firebase... following snippet works:
`YOUR_FIREBASE_TOKEN_STREAM() => auth.FirebaseAuth.instance.idTokenChanges().listen((auth.User? user) async {
  if (user != null) {
      final token = await user.getIdToken();
      // We maintain the Supabase profile id in Firebase so we pull that out and send it to the Token minting function so that it can use
      // info as additional claim - you may not need this.  In that case you can ignore the 3rd param of supabaseId
      final supaToken = await mintTokenFor(token, user.uid, supabaseId);
  }
});

// Next call the Supabase function and consume the response
  final response = await Supabase.instance.client.functions.invoke('mint-token',
      body: {
        'token': token,
        'firebaseUser': {'uid': uid, 'supabase_id': supabaseId},
      });
    Supabase.instance.client.headers = {
    'Content-Type': 'application/json',
    'Authorization': 'Bearer ' + supaToken,
  };
  Supabase.instance.client.storage.setAuth(supaToken);
  Supabase.instance.client.rest.setAuth(supaToken);
  Supabase.instance.client.realtime.setAuth(supaToken);
  Supabase.instance.client.functions.setAuth(supaToken);
`

**Comment by dhurbhat** [comment]
@evelant - this is awesome! Thanks!
QQ: the JWT secret is the Supabase JWT secret - correct?
(I think it has to be as RLS would need to use this secret to verify the newly minted JWT).

Also - I your RLS policy, the "userId" is a column on `activities` correct?

**Reply by evelant** [reply]
Yes and yes

**Reply by evelant** [reply]
In porting from firebase I had lots of case sensitive column names which is not how things are usually done in postgres. You have to double quote identifiers in postgres otherwise they're treated as all lower case.

**Comment by dhurbhat** [comment]
Subhubirl,
If you have git access to code generated by flutter flow, then you can use
this - you only get access to git of you have paid ff subscription. If you
are on free version you won’t be able to use this as it requires
manipulating the headers to supabase invocations…

**Reply by subhubirl** [reply]
oh wow - yes I have a paid subscription to FF. Would you recommend to just follow [these steps](https://docs.flutterflow.io/customizing-your-app/manage-custom-code-in-github)? 
Also, does anything need to be edited in the above code - references to my environment etc - or is it good to be copied over directly? Been searching for this answer for days now, would be unbelievable if I can get it to work

**Comment by Darren-Ter** [comment]
Hey @evelant, big thanks for your guide, it's been a real lifesaver. Just got 2 things I'm curious about:

1. How do you know when it's time to re-generate Supabase Token? With Firebase, their ID Tokens last only an hour, and they've got this [onIdTokenChanged](https://firebase.google.com/docs/reference/js/auth.md#onidtokenchanged_b0d07ab) listener to keep track of updates. But from what I've tested, it's not always reliable – sometimes it doesn’t trigger the action in listener when the ID Token refreshes. I'm curious is there a more reliable method to trigger the edge function to re-generate Supabase token every time the Firebase ID Token refreshes?

2. About the edge function to re-generate Supabase token, does it mean I need to trigger it around 720 times a month for each user (24 * 30, as Firebase ID Token refreshes every hour)? I mean, Supabase has a pretty cool free plan, but I’m starting to think this could get pricey with a lot of users. Can you share any insights from your experience about how this might affect costs?

Would really appreciate your thoughts on this. Thank you!

**Reply by evelant** [reply]
I use the `getIdTokenResult(false)` method to get the decoded token details from the firebase sdk. Whenever my client is going to make a call I check if it (a cached copy of the token) is going to be expired soon and refresh if necessary. Ex: ` firebaseTokenDetails.token.claims.exp * 1000 < Date.now() - 1000 * 60 * 10` to check if the token expires within the next 10 minutes. If it expires soon then I call `getIdTokenResult(true)` to force firebase to refresh the token and I use the refreshed token to call my edge function for a new supabase token.

As for the potential costs, I'm not sure, it depends on your use case. I actually don't use supabase edge functions anymore. It's the one part of the stack I don't like. Deno causes way too many headaches. It was way too hard to get shared code to run in both Deno and my app. Deno's vscode plugin is buggy and broken in a lot of ways so type checking never worked (especially in a monorepo). Edge functions also failed at random, with ~5% of requests never reaching my code. As an alternative I opted to deploy a simple Bun server on fly.io. It's not any harder than deploying supabase edge functions but it's vastly more reliable and all the tooling just works because no Deno. Never touching Deno again if I can avoid it lol

**Reply by Darren-Ter** [reply]
Thanks for sharing your method and experiences! The approach of checking for token expiry and refreshing it as needed seems like a smart move.

I was actually planning to migrate all my APIs to Supabase edge functions since my user base is spread across Canada, Australia, Southeast Asia, and Taiwan (and potentially US & Europe in future). I thought it would be a good fit, considering the geographical spread. But hearing about your struggles with Deno and the reliability issues you faced is making me rethink that strategy 😂. I'll take your advice to heart and will closely monitor my edge functions before fully deploying them

**Reply by evelant** [reply]
The Supabase team is awesome and has made great progress in improving the edge-functions so they're worth a shot if Deno doesn't pose any big problems for you. If you don't have shared code between an app and edge functions and don't use a monorepo then Deno shouldn't be too much trouble especially now that it can use npm modules.

IMO Deno as a project just went down the wrong path, it tried to invent its own "better" way of doing things that was incompatible with the larger ts ecosystem and has been slowly backtracking into compatibility since (like adding npm support). I think Bun really gets it right where Deno missed the mark (I wish the supabase team would adopt it for edge-runtime instead of Deno). Bun is super fast, works with all modern js/ts code, and includes all the tooling you need to develop a project. Saves time and much nicer than cobbling together all the parts from various slightly incongruent tools from npm. 

BTW if supabase edge functions don't work out for you I've had a great experience with https://fly.io and I think it's a perfect alternative. Fly can do the same "deploy near users" trick that supabase edge functions do. It was really easy to set up. I was kicking myself for spending so much time fighting with Deno after I got all my code running in bun and deployed to fly in just a couple hours.

**Comment by technopahadi** [comment]
@evelant's answer works pretty good.

One thing I wanted to mention regarding RLS, if you do not want to set the ID in `sub` property but use a different one, then in order to access that you can use the following:

```
-- Note `claims` and not `claim` (plural vs singular)
current_setting('request.jwt.claims', false)::json->>'PROPERTY');
```

**Comment by christiangenco** [comment]
I solved this by writing a Firebase callable function that: 

- checks that the Firebase user is authenticated in `request.auth`
- pulls the Firebase user info from my Firestore `users` table (but this could also come from the Firebase auth)
- upserts the user into Supabase with [`supabase.auth.signUp`](https://supabase.com/docs/reference/javascript/auth-signup), including populating a `public.profiles` table
- generates a one time password (OTP) hashed token with [`await supabase.auth.admin.generateLink({ type: "magiclink", email})`](https://supabase.com/docs/reference/javascript/auth-verifyotp)
- returns the token to the client

Then on my client I wrote a hook that:

- checks if I'm already logged into Supabase by looking for a cookie named `sb-PROJECTID-auth-token` (but this doesn't feel like the right way to check if Supabase is already logged in)
- if I'm not logged into Supabase *and* I'm logged in with Firebase, then the hook calls that callable function from above to fetch the OTP `hashed_token`
- authenticates with Supabase with [`supabase.auth.verifyOtp({ token_hash: properties.hashed_token, type: "email" })`](https://supabase.com/docs/reference/javascript/auth-verifyotp)

Now I can make a super smooth transition from Firebase to Supabase because if I have a logged in Firebase user I also have a logged in Supabase user.

The next stage of my migration is to add `onDocumentWritten` triggers for each of my collections that will transform each Firestore document and insert it into Supabase because the standard [Firestore Data Migration](https://supabase.com/docs/guides/resources/migrating-to-supabase/firestore-data) would require some downtime.

**Comment by aplyd** [comment]
@christiangenco I'm currently doing something similar except we didn't want to keep our Firebase project active until all users fully migrated (potentially indefinitely).

Instead, I'm using Supabase's Firebase foreign data wrapper to migrate Firebase Auth and Firestore Collections data into Supabase. From the Firebase tables I've created a materialized view that is refreshed via pg_cron (but will stop being refreshed after we fully remove Firebase Auth from all of our clients).

When a user login attempt fails, we check our migrated Firebase Auth data for a matching email - if there's a matching email we then hash the password entered during the failed login attempt via `firebase-scrypt`, compare the hashes, and if they match we create a new user in Supabase Auth with the same credentials. And finally, I have an `auth.users` insert trigger to take any migrated users data from the Firebase materialized view and insert it into the users new Supabase records.

The password hashing is done in a cloud function but the remaining logic happens mostly in the database. Just wish I could somehow use the `firebase-scrypt` package within Postgres to remove the need for a cloud function at all.

I'd be curious to hear anyone's feedback on this approach.

[Firebase Scrypt](https://github.com/firebase/scrypt) (I used the Node implementation [here](https://github.com/xeewi/firebase-scrypt) in my cloud function)
